### ISSUE TITLE
Allow sharing of individual windows.

### ISSUE THREAD

**Issue by selflein** [emersion/xdg-desktop-portal-wlr#107]
Adding the option to share individual windows was discussed in different issues/PRs but there is no corresponding open issue right now.

**Comment by selflein** [comment]
Some discussion can be found [here](https://github.com/emersion/xdg-desktop-portal-wlr/issues/12#issuecomment-770377796)

**Comment by danshick** [comment]
True, thanks. Also worth linking to the [corresponding wlr-protocols issue](https://github.com/swaywm/wlr-protocols/issues/93), which I would consider a prerequisite/blocker for this feature.

**Comment by timsofteng** [comment]
Hello. What is a current state of this feature?

**Comment by hugopeixoto** [comment]
Here's the link to the same issue mentioned by danshick in the new forge: https://gitlab.freedesktop.org/wlroots/wlr-protocols/-/issues/93

**Comment by tm512** [comment]
I'd really like to see a feature like this implemented. My streaming setup on X11 involves having a game in a window on one side of the screen with stuff like terminals and twitch's dashboard on the other side of the screen. Since I'm stuck with a single-monitor setup, I don't have the option of just running the game fullscreen on one monitor and using fullscreen capture, with everything else on the other monitor.

It sounds like some other Wayland compositors like GNOME already provide the ability for single-window capture? Maybe I'm misunderstanding, but if that's the case then what are they doing that makes that possible that wlroots doesn't do?

**Comment by tm512** [comment]
>For the meantime there is https://github.com/emersion/xdg-desktop-portal-wlr/pull/156 to share a region of the screen.

Unfortunately specifying the cropping region ahead of time isn't really practical without a way to precisely move and resize a window into that region (like what `xdotool` can easily do on X11). While things are limited to capturing entire outputs, would it be possible to run the game to be captured inside of a nested wayland compositor running in a window? I'm not sure if its buffer would then be available for capture via xdpw. The ability to set the resolution of the nested compositor would also be important in that case.

If that's not feasible, I've been thinking of just hacking whatever compositor I settle on to keep a tmpfile with a full list of windows and their size/position, then pass those into `slurp` to interactively select which window to capture, using that with wf-recorder or ffmpeg w/ kmsgrab.

**Comment by wallabra** [comment]
> If that's not feasible, I've been thinking of just hacking whatever compositor I settle on to keep a tmpfile with a full list of windows and their size/position, then pass those into `slurp` to interactively select which window to capture, using that with wf-recorder or ffmpeg w/ kmsgrab.

@tm512  Another concern is that a window is not always in the foreground, like when using workspaces or a tabbed container in Sway.

**Comment by spagootie** [comment]
This feature would be great to have, because it's on virtually every platform other than wlroots compositors.

**Comment by sfrique** [comment]
It's bad, but it's able to share only "a few things" if you need.  For me it solved the biggest problem, not wanting to show my full output. 

it seems to work just "fine", but did not test it enough.

You will need to run a nested Wayland session.. i tested with sway  what i did to make it work:

1. Created a clean sway config (no auto start bar etc...) run `sway -c ..../sway-nested`
2. Created a pass-through mode on my main sway (to be able to run any command on my nested sway)
3. Created a chromium/firefox profile (just so i don't have to close my current one at the "main" sway)
4. Killed all xdg portal (if i did not do this, it would share the whole output, even when at nested session)
5. Share the output, since this is nested, it only shared the nested session, that on my test was a floating windows of sway with whenever opened there


as I said, it's ugly.. I just did some initial testing, did not check audio etc etc yet. 
I was just trying to solve a problem I now have, since i am moving to a single big monitor, i wasn't able to share my second output and keep doing stuff while sharing the screen, but this should solve this.. since nothing would be running at nested, only what i actually want.


I don't know if nested session work with all the apps and etc.. but if it's only one instance of it, and it's running in the nested, should be fine.

There is a lot of improvement and tests that can be made, i just wanted to share something i just got working.

Hope that helps people =]

**Comment by wallabra** [comment]
You see, this nesting should naturally derive from "surfaces" as a first-class abstraction of rendered regions, indifferent to whether it's a desktop, a window, or whatever else. A tree of surfaces should be enumerable and visible from the top, globally but probably with some sort of access control scheme. The issue is this was not considered initially in the design of _wlroots_, which will haunt it for a very long time because of a little something called [technical debt](https://en.wikipedia.org/wiki/Technical_debt).

**Comment by cmprmsd** [comment]
I wonder why nested Sway is a functioning workaround but it's not possible to share individual windows. For now I just share via xwayland, which works okay and is more easy to use than having nested WMs.

**Comment by ckcr4lyf** [comment]
> For now I just share via xwayland

@cmprmsd can you expand on this - do you mean the specific window you want to share is what you launch via xwayland?

**Comment by cmprmsd** [comment]
Sure thing. 
When you explicitly launch two applications like chromium and e.g. VMware workstation with xwayland, than both will use "the old" methods for sharing and therefore see each other. 

In Chromium you'll notice that for example in Teams you will see all xwayland windows. The monitor share tabs will show black screens. But if you switch to single window, than you can select windows started under xwayland. 

Hope this helps for the time being.

**Comment by WhyNotHugo** [comment]
Following up on the [roadmap above](https://github.com/emersion/xdg-desktop-portal-wlr/issues/107#issuecomment-1056817342):

> - Create a way to export handles for toplevels (windows) from the compositor

This will be covered by [ext-foreign-toplevel-info](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/187).

**Comment by cmprmsd** [comment]
My workaround does not work any more since some weeks. I don't know why but some component been updated and now only the "Wayland-dialog" is visible even for Chromium or Firefox which have been started with xWayland instead of Wayland.  :/

I looked at the pipewire settings and some other flags but nothing did work. :cry:

**Comment by jalvesaq** [comment]
A workaround for this is to create a virtual output (`HEADLESS-1`), move the window to it, and share the virtual output. My configuration for this is:

`~/.config/xdg-desktop-portal-wlr/config`:

```
[screencast]
chooser_type=dmenu
chooser_cmd=swaymsg -t get_outputs | jq '.[] | .name' | sed 's/"//g' | wofi -d
max_fps=30
```

`~/.config/sway/config`:

```
# Do the following command in a terminal emulator when you need the virtual output:
# swaymsg create_output

output HEADLESS-1 resolution 1280x720 position 0,1080
output HEADLESS-1 bg "#220900" solid_color
workspace 0 output HEADLESS-1
bindsym $mod+0 workspace number 0
bindsym $mod+Shift+0 move container to workspace number 0
```

The command `swaymsg create_output` creates the virtual output `HEADLESS-1` the first time that it is run. If you repeat the command, it will create the virtual outputs `HEADLESS-2`, `HEADLESS-3`, and so on...

By default, when I try to share an output, the cursor turns a cross (`slurp -o` is running) and I have to click anywhere in the output to choose it. I could move the cursor to the virtual output, but I prefer to see the list of available outputs. That's why I configured `xdg-desktop-portal-wlr` to use `wofi` instead of `slurp`.

**Comment by clushie** [comment]
@jalvesaq Oh cool, so the idea is to overlay a headless output (what you are doing by picking the right pixel position on creation)  with a regular output and then to share that headless output? I just tested it quickly by using `wdisplay` to move the headless output on the same screen as I'm currently at, unfortunately only shows up if I set it to floating and place it within the positioning of the headless output. How do you solve this? PS: You can do `jq -r` to remove the `"` symbols.

**Comment by trinitronx** [comment]
> I can use `wl-mirror` to visualize the headless output

Very nice!  I'll have to try this method out.

> But I can also see it in Google Meet (if I'm sharing the output) or preview it in OBS as a "Screen Capture (PipeWire)" source.

Another way that I've found is to use `wayvnc` + VNC client to view the headless output:

    wayvnc --gpu --performance --output=HEADLESS-1
    # Then connect to 127.0.0.1:5900 with VNC client (e.g. Remmina)

Keyboard capture becomes somewhat of an obstacle with this method, but usually I can switch back out of the `HEADLESS-*` output VNC window using the Sway workspace switching keyboard shortcuts.  Also, unfortunately the VNC "Host key" doesn't work well (default in Remmina is `Ctrl_R` / right <kbd>Ctrl</kbd>).

**Comment by paolomainardi** [comment]
I am curious to know if other portals have found a solution to this problem?

**Comment by WhyNotHugo** [comment]
@paolomainardi Hyperland uses the [`hyprland-toplevel-export-v1`](https://github.com/hyprwm/hyprland-protocols/blob/main/protocols/hyprland-toplevel-export-v1.xml) protocol, which wlroots does not support.

---

Based on [the roadmap above](https://github.com/emersion/xdg-desktop-portal-wlr/issues/107#issuecomment-1056817342), the current state is:

> Finish [ext-screencopy](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/124)

Still in progress, lots of unsolved discussion there.

> Create a way to export handles for toplevels (windows) from the compositor

Done. [Implemented via the ext-foreign-toplevel-list protocol](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/187).

> Expand ext-screencopy to accept these handles

This could be implemented experimentally in zwlr_screencopy_manager_v1. Although I'd ask the wlroots devs if they'd merge such a change at this stage or would rather wait for ext-screencopy.

> Expand the chooser protocol to account for them (https://github.com/emersion/xdg-desktop-portal-wlr/issues/169)

The [OUTPUT CHOOSER](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/xdg-desktop-portal-wlr.5.scd#output-chooser) protocol can be extended to take a value matching a [ext_foreign_toplevel_handle_v1::identifier](https://wayland.app/protocols/ext-foreign-toplevel-list-v1#ext_foreign_toplevel_handle_v1:event:identifier).

Given that, potentially, and output name can match a toplevel name (it would be terrible, but nothing explicitly forbids this), I think a prefix of `toplevel:` should be used here.

> Add this together to share window buffers

Blocked by the items above

**Comment by paolomainardi** [comment]
Wow, thanks, @WhyNotHugo. That's a great summary. So we have to help to push `ext-screenshopy` further.

**Comment by amano-kenji** [comment]
Is hyprland already more mature than sway? I already found a few things that work properly only in hyprland.

**Comment by RioMeYeah** [comment]
Still nothing with Sway?

**Comment by ngotchac** [comment]
I have this little script to start a new headless output for screen sharing:
```bash
#!/usr/bin/env bash

# Step 1: Create a new output
swaymsg create_output

# Step 2: Find the name of the newly created output
NEW_OUTPUT=$(swaymsg -t get_outputs | jq -r '.[] | select(.name | startswith("HEADLESS-")) | .name' | sort | tail -n 1)

# Check if the output was successfully created
if [ -z "$NEW_OUTPUT" ]; then
    echo "Failed to create a new output."
    exit 1
fi

# Step 3: Assign a workspace to the new output
swaymsg workspace sshr output "$NEW_OUTPUT"

# Step 4: Set the resolution for the new output
swaymsg output "$NEW_OUTPUT" resolution 1280x720

# Step 5: Set the background color for the new output
swaymsg output "$NEW_OUTPUT" bg "#220900" solid_color

# Step 6: Switch to workspace sshr and then back to the previous workspace
CURRENT_WORKSPACE=$(swaymsg -t get_workspaces | jq -r '.[] | select(.focused) | .name')
swaymsg workspace sshr
swaymsg workspace "$CURRENT_WORKSPACE"

wl-mirror "$NEW_OUTPUT"

notify-send "Created new output $NEW_OUTPUT."
```

with keybindings in my sway config for `workspace sshr`.

It creates the new output, assign the `sshr` workspace to it, switch back and forth to it (since when the output is created, the next available workspace is assigned to it), open `wl-mirror` to it, and send a notification.

**Comment by rmasad** [comment]
Nice, the problem is that now Chrome doesn't let me share windows, only tabs. Pressing on windows returns me to the tabs section.

**Comment by jalvesaq** [comment]
@rmasad, you can share the whole virtual (headless) output and use `wl-mirror` only to visualize the output outside the browser.

**Comment by YellowOnion** [comment]
> @paolomainardi Hyperland uses the [`hyprland-toplevel-export-v1`](https://github.com/hyprwm/hyprland-protocols/blob/main/protocols/hyprland-toplevel-export-v1.xml) protocol, which wlroots does not support.

Hyprland uses a fork of wlroots, the question of why they haven't migrated these changes to upstream is valid, though I think just a boring answer where they just want to move fast and break things without worrying about compatibility till designs are stable or trying be considerate of ever potential use case, I also have a suspicion that wlroots/wayland devs have other priorities, hyprland has a focus on features that make using OBS good, like global hotkey support, that's been a stagnant in wayland for about 8 years, and hyprland is the only wl-wm with support for it.

**Comment by PhpOfc** [comment]
> I have this little script to start a new headless output for screen sharing:
> 
> ```shell
> #!/usr/bin/env bash
> 
> # Step 1: Create a new output
> swaymsg create_output
> 
> # Step 2: Find the name of the newly created output
> NEW_OUTPUT=$(swaymsg -t get_outputs | jq -r '.[] | select(.name | startswith("HEADLESS-")) | .name' | sort | tail -n 1)
> 
> # Check if the output was successfully created
> if [ -z "$NEW_OUTPUT" ]; then
>     echo "Failed to create a new output."
>     exit 1
> fi
> 
> # Step 3: Assign a workspace to the new output
> swaymsg workspace sshr output "$NEW_OUTPUT"
> 
> # Step 4: Set the resolution for the new output
> swaymsg output "$NEW_OUTPUT" resolution 1280x720
> 
> # Step 5: Set the background color for the new output
> swaymsg output "$NEW_OUTPUT" bg "#220900" solid_color
> 
> # Step 6: Switch to workspace sshr and then back to the previous workspace
> CURRENT_WORKSPACE=$(swaymsg -t get_workspaces | jq -r '.[] | select(.focused) | .name')
> swaymsg workspace sshr
> swaymsg workspace "$CURRENT_WORKSPACE"
> 
> wl-mirror "$NEW_OUTPUT"
> 
> notify-send "Created new output $NEW_OUTPUT."
> ```

A awesome script! just one question, how do i delete the virtual output (the "headless" one) after i'm done recording/sharing screen?

**Comment by ngotchac** [comment]
> A awesome script! just one question, how do i delete the virtual output (the "headless" one) after i'm done recording/sharing screen?

I use this script for stopping the screenshare:
```bash
#!/usr/bin/env bash

# Get all outputs with names starting with HEADLESS-
headless_outputs=$(swaymsg -t get_outputs | jq -r '.[] | select(.name | startswith("HEADLESS-")) | .name')

# Check if there are any HEADLESS outputs
if [ -z "$headless_outputs" ]; then
    echo "No HEADLESS outputs found."
    exit 0
fi

# Unplug each HEADLESS output
for output in $headless_outputs; do
    echo "Unplugging $output..."
    swaymsg output "$output" unplug
done

notify-send "All HEADLESS outputs have been unplugged."
```

**Comment by pm4rcin** [comment]
@WhyNotHugo so now things just need to be implemented in wlroots and integrate OUTPUT CHOOSER and finally there's a generic way to share windows on wlroots based compositors right? :)

**Comment by WhyNotHugo** [comment]
Indeed. An updated version of the [previous roadmap](https://github.com/emersion/xdg-desktop-portal-wlr/issues/107#issuecomment-1860051956) is:

- wlroots implementation (https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4545)
  - compositor implementation for your favourite compositor
- Update the [OUTPUT CHOOSER](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/xdg-desktop-portal-wlr.5.scd#output-chooser) protocol to take a [ext_foreign_toplevel_handle_v1::identifier](https://wayland.app/protocols/ext-foreign-toplevel-list-v1#ext_foreign_toplevel_handle_v1:event:identifier).

**Comment by llyyr** [comment]
> wlroots implementation (https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4545)

This doesn't actually implement [ext_foreign_toplevel_image_capture_source_manager_v1](https://wayland.app/protocols/ext-image-capture-source-v1#ext_foreign_toplevel_image_capture_source_manager_v1) so that's not sufficient btw.

**Comment by pranavsetpal** [comment]
> wlroots implementation (https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4545)

this has been merged! @llyyr pointed out this is not enough, but still good news :)

**Comment by paolomainardi** [comment]
Finally, I cannot wait to have the feature implemented, not being able to share windows on calls is a big downside compared to the other major window managers, including Hyprland.

**Comment by eternal-sorrow** [comment]
>This doesn't actually implement [ext_foreign_toplevel_image_capture_source_manager_v1](https://wayland.app/protocols/ext-image-capture-source-v1#ext_foreign_toplevel_image_capture_source_manager_v1) so that's not sufficient btw.

Where is this protocol coming from? Does it even exist in wayland-protocols?

**Comment by hvhaugwitz** [comment]
What is the state of this feature request? Is there an updated roadmap with the still missing features?

**Comment by heroin-moose** [comment]
Somehow Hyprland's portal have been allowing this for years, even before ditching wlroots. Maybe worth taking a look at what vaxry did?

**Comment by llyyr** [comment]
> Somehow Hyprland's portal have been allowing this for years, even before ditching wlroots. Maybe worth taking a look at what vaxry did?

They did so with a private protocol, but there's no need for that now because ext-image-capture-source-v1 and ext-image-copy-capture-v1 exist. It just needs someone to implement it in wlroots and then in this portal.

**Comment by llyyr** [comment]
> Which probably already is done in the RC releases for sway 1.11 and wlroots 0.19 by reading these release notes: https://github.com/swaywm/sway/releases/tag/1.11-rc1

No, as I've mentioned before the wlroots implementation doesn't support [ext_foreign_toplevel_image_capture_source_manager_v1](https://wayland.app/protocols/ext-image-capture-source-v1#ext_foreign_toplevel_image_capture_source_manager_v1) yet

**Comment by paolomainardi** [comment]
related https://github.com/swaywm/sway/issues/8692

**Comment by heroin-moose** [comment]
Wayland at it's finest, LOL.

**Comment by RangerMauve** [comment]
This would be very useful to have 😅

**Comment by wallabra** [comment]
> Wayland at it's finest, LOL.

It's just a heck of a lot of effort to make a whole family of composable protocols and make them work together, especially when you're trying to **not** immediately break everything people have ever made, every single time you want to update a protocol or add a new one.

**Comment by tmccombs** [comment]
It's a social issue, not a technical problem. There are several parties involved that all want different things from Wayland, and getting them all to agree on a protocol is difficult.  And then of course, once a protocol is decided on, it takes time for all the applicable implementations to actually implement it.

**Comment by mvdan** [comment]
Please take this discussion elsewhere; dozens of us are subscribed to this thread for updates on the task at hand, and you're quickly getting off-topic.

**Comment by karras** [comment]
Isn't https://github.com/emersion/xdg-desktop-portal-wlr/pull/329 working towards sharing of individual windows?

**Comment by wallabra** [comment]
> Isn't [#329](https://github.com/emersion/xdg-desktop-portal-wlr/pull/329) working towards sharing of individual windows?

Yes and no. It has not received any updates in two weeks, and no commits in a month already. It's only relevant if the original author of the PR returns, or someone else takes the torch.

**Comment by paolomainardi** [comment]
Things are moving, here some updates: https://emersion.fr/blog/2025/status-update-77/

**Comment by wallabra** [comment]
Yes, and that is nice! Note that that was in June. I am excited to see where this goes.

**Comment by cmprmsd** [comment]
Stop discussions please. This will not move things forward. As mentioned above there are many people subscribed to this issue for updates. Not discussions.

**Comment by zukigay** [comment]
Using sway git with wlroots git and xdg-desktop-portal-wlr git and finally have working screen share of single windows working perfectly in firefox

**Comment by paolomainardi** [comment]
Thanks a lot, you all rock.

**Comment by wallabra** [comment]
> Using sway git with wlroots git and xdg-desktop-portal-wlr git and finally have working screen share of single windows working perfectly in firefox

Whoa! I assume this means this is coming to other wlroots compositors, like niri, as well, right? I am so excited!

**Comment by diniamo** [comment]
Niri is not wlroots, but it's already supported this before, from what I can tell.

**Comment by karras** [comment]
@emersion Is another release planned in the foreseeable future to include the new screen sharing capabilities?

**Comment by emersion** [comment]
We don't do ETAs, but we'll ship a new release with this feature yes.

**Comment by amano-kenji** [comment]
Which issue tracks slurp window chooser?

**Comment by arsnazarenko** [comment]
> This is something I've noticed a while ago, and it's not supposed to happen. Can you open a Sway bug about this?

https://github.com/swaywm/sway/issues/8885

**Comment by emersion** [comment]
I can cut a new xdpw release easily, but the latest wlroots (and Sway) releases don't ship support for toplevel capture, so not sure if that would be useful to you still?

**Comment by guille** [comment]
For those wondering why window sharing still doesn't work on the latest tag, as said above

> the latest wlroots (and Sway) releases don't ship support for toplevel capture

So it needs [Sway 1.12](https://github.com/swaywm/sway/issues/8754) to be released

**Comment by Stachelbeere1248** [comment]
is there any ETA for wlroots 0.20? bc afaik, thats required for the compositors to release this feature, or at least they said so in a river-classic pr.

btw been using the branches of said pr and feature has been working nicely.
Might be a Discord issue and maybe I'm asking in the wrong place here but it always asks for the output three times instead of just once, which is a minor annoyance.
